presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java

```diff
@@ -51,264 +43,104 @@ private RegexpFunctions()
 
     @ScalarOperator(OperatorType.CAST)
     @SqlType(RegexpType.NAME)
-    public static Regex castToRegexp(@SqlType(StandardTypes.VARCHAR) Slice pattern)
+    public static RegexpGenericPattern castToRegexp(ConnectorSession session, @SqlType(StandardTypes.VARCHAR) Slice pattern)
     {
-        Regex regex;
+        String regexLibraryName = session.getSystemProperty(REGEX_LIBRARY, String.class);
+        RegexLibrary regexLibrary;
+        if (regexLibraryName.toUpperCase().equals("RE2J")) {
+            regexLibrary = RE2J;
+        }
+        else if (regexLibraryName.toUpperCase().equals("JONI")) {
+            regexLibrary = JONI;
+        }
+        else {
+            throw new PrestoException(INVALID_SESSION_PROPERTY, "The setting for '" + REGEX_LIBRARY + "' is not valid. It should be either RE2J or JONI.");
+        }
+
+        int dfaStatesLimit = session.getSystemProperty(RE2J_DFA_STATES_LIMIT, Integer.class);
```

Comment:
The states limit is used as a rough form of memory management.  RE2J has two algorithms that it can use-- DFA and NFA.  Both grow linearly with the size of the input, but DFA is substantially faster.  However, the DFA algorithm can use a ton of memory for certain inputs and pattern combinations.  Therefore we allow configuration of a state limit.  Once it hits the states limit, the re2j will fallback to using the NFA algorithm on that input.

The retry limit defines how many times re2j hits the states limit before running all of the inputs for a pattern on the nfa.  The reason for this is that whenever re2j needs to fallback to the nfa algorithm, it has to create a new dfa machine for the rest of its inputs, which is expensive if every input is going to have to fall back to nfa anyway.  However, if the inputs that hit that limit are outliers, then it's worth the cost because of the performance benefits of using the DFA on the other inputs.   You might want a higher retry limit if you are going to be running on larger datasets.